man/unicoverage.1
.TH UNICOVERAGE 1 "2007 Dec 31"
.SH NAME
unicoverage \- Print coverage of each Unicode BMP Script
.SH SYNOPSIS
.br
.B unicoverage
[-iinput-file] [-ooutput-file]
.SH DESCRIPTION
.B unicoverage
reads a GNU Unifont .hex font and uses data in
.B coverage.dat
(which must reside in the current directory).  The output is
the percent coverage of each script in the Unicode Basic Multilingual Plane.
.SH OPTIONS
.TP 12
-i
Specify the input file. The default is STDIN.
.TP
-o
Specify the output file. The default is STDOUT.
Sample usage:
.PP
.RS
unicoverage < unifont.hex >coverage.txt
.RE
.SH FILES
coverage.dat, *.hex GNU Unifont files.
.SH SEE ALSO
.BR bdfimplode(1),
.BR hex2bdf(1),
.BR hex2sfd(1),
.BR hexbraille(1),
.BR hexdraw(1),
.BR hexkinya(1),
.BR hexmerge(1),
.BR johab2ucs2(1),
.BR unibdf2hex(1),
.BR unibmp2hex(1),
.BR unidup(1),
.BR unifont(5),
.BR unifont-viewer(1),
.BR unifont1per(1),
.BR unifontchojung(1),
.BR unifontksx(1),
.BR unifontpic(1),
.BR unigencircles(1),
.BR unigenwidth(1),
.BR unihex2bmp(1),
.BR unihex2png(1),
.BR unihexfill(1),
.BR unihexgen(1),
.BR unipagecount(1),
.BR unipng2hex(1)
.SH AUTHOR
.B unicoverage
was written by Paul Hardy.
.SH LICENSE
.B unicoverage
is Copyright \(co 2007, 2008 Paul Hardy.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.SH BUGS
No known real bugs exist, except that this software does not perform
extensive error checking on its input files.


NEWS
2017-06-20: Release 10.0.01

   Updated for Unicode 10.0 release.

   New scripts added (introduced in Unicode 10.0):
     - U+0860..U+086F     Syriac Supplement
     - U+011A00..U+011A4F Zanabazar Square
     - U+011A50..U+011AAF Soyombo
     - U+011D00..U+011D5F Masaram Gondi

   Added quadruple-width glyphs U+9FD8..U+9FE9, updating some utility
   programs and documentation to allow this change.  The Perl scripts
   unihex2png, unipng2hex, and unifont-viewer do not support this new
   16-by-32 glyph size.

   Corrected several glyphs and made numerous changes in others,
   described in detail in the ChangeLog file.

2016-12-22: Release 9.0.06

   Refined about 50 Pikto CSUR glyphs.

   Corrected two Osage glyphs and one mathematical operator as noted
   in ChangeLog.

2016-12-10: Release 9.0.05

   Modified Cherokee glyphs to reflect changes in recent Cherokee font
   that the Unicode Consortium adopted for its code charts.  Removed
   serifs to make Cherokee glyphs easier to read.

   Modified Pikto CSUR glyphs after comparison with high-resolution scans
   of Basic Pikto book.

2016-10-29: Release 9.0.04

   Extended "{" and "}" by one pixel so they now come to a point in the center.
   Swapped U+2A05 and U+2A06 so they are now in the correct order.

2016-10-21: Release 9.0.03

   Added CSUR glyphs:
     - Aiha: added doubling combining glyph, omitted from the ConScript
       Unicode Registry (CSUR) code chart.
     - Tonal System: base 16 counting system devised by John William Nystrom
       in the mid-19th century.
     - Pikto: added the Pikto glyphs to Plane 15 Private Use Area, in the
       range specified in the CSUR.

   The giant picture of all of Plane 0 now includes the Plane 0 CSUR glyphs,
   although those are an optional part of Unifont in separate font files.

2016-08-27: Release 9.0.02

   Added 3 more CSUR scripts.  Modifications to some existing glyphs.

2016-07-02: Release 9.0.01

   Updates to support Unicode 9.0.0.

2015-06-28: Release 8.0.01

  Updates to support Unicode 8.0.0.

2014-10-23: Release 7.0.06

   A few new scripts were added:
      U+010350..U+01037F Old Permic
      U+01F650..U+01F67F Ornamental Dingbats
      U+01F780..U+01F7FF Geometrical Shapes Extended
      U+01F800..U+01F8FF Supplemental Arrows-C

   Various changes were made to Makefiles, Perl scripts, and one bash
   shell script to conform to Debian Policy Manual version 3.9.6.

   Several changes were made to Makefiles:
      * Use "set -e" at the start of multi-command shell invocations,
        to trap errors.
      * Use "&&" rather than ";" in multi-command shell invocations,
        to fail on errors.
      * install now uses "-s" flag for C program binaries.

   Other minor changes were made to scripts; see ChangeLog
   for details.

2014-10-17: Release 7.0.05

   Modified ASCII letter 'l' so it would have the same height as
   the other lower-case letters with ascenders; this is one pixel
   taller than the capital letters.  Lowered the horizontal stroke 
   in 't' by one pixel so it would line up with the horizontal
   stroke in 'f'; this is one pixel lower than the x-height of
   the ASCII letters.  Letters elsewhere in Plane 0 that were formed
   with 'k', 'l', or 't' were checked and modified where appropriate
   to conform with the new ASCII glyphs.

   Completed the Miscellaneous Symbols and Pictographs script in Plane 1,
   and modified Plane 1 Emoticons and Transport and Map Symbols scripts.

   Now create file font/compiled/plane01.html, showing a color-coded
   indication of completion of all of Unicode Plane 1 as an HTML
   table in a self-contained HTML file.  That and the compiled/index.html
   are the two HTML files that building the font creates.  The plane01.html
   file gives a color coding to show completion of each 256-glyph block,
   from 0% (pure red) to 100% (pure green).

2014-10-11: Release 7.0.04

   Added five more Supplemental Multilingual Plane (Plane 1) scripts
   and improved others thanks to several contributors.  There are now
   over 5,000 drawn assigned glyphs for Plane 1!

   Reorganized font/plane01 directory.  Several other changes; see
   ChangeLog file for details.

   unihex2sfd and unipagecount now handle all Unicode planes, not just
   Plane 0.


2014-07-01: Release 7.0.03

   Minor release for Debian.  Added a man page for unihexfill(1);
   Debian requires a man page for each program.

2014-06-22: Release 7.0.02

   Fixes a bug in font/Makefile: there were two "done" statements at the
   end of one "for" loop.  Also adds many new Supplemental Multilingual
   Plane (Plane 1) scripts.  See the ChangeLog file for details.

2014-06-21: Release 7.0.01

   Provides a glyph for every visible code point in the Unicode 7.0 Basic
   Mulitlingual Plane.  Also adds more scripts in the Unicode 7.0
   Supplemental Multilingual Plane.  See the ChangeLog file for full
   details.

   Note that the versioning scheme has changed.  Instead of a date, this
   package will use an incrementing minor version number in the future
   ("01", "02", etc.).

2014-02-14: Release 6.3.20140214

   This version adds the GNU Free Documentation License 1.3 to the "COPYING"
   file.

2014-02-04: Release 6.3.20140204

   This version added Qianqian Fang's name to the COPYRIGHT string now
   passed to Fontforge when the TTF fonts are built, for his contributions
   contained in the file "font/plane00/wqy.hex".

   This version also removes "<img>" links to icons on external websites from
   the "hangul/hangul-generation.html" web page, which were retrieved and
   stored locally when the original web page was saved in the "hangul/"
   directory.

2014-02-02: Release 6.3.20140202

   This version is the first designed to easily build multiple fonts.
   To accomplish this, font/Makefile and font/ttfsrc/Makefile were
   modified to pass FONT and COMBINING parameters, giving the names
   of the font file and its associated combining.txt file.


src/unicoverage.c
/*
   unicoverage - Show the coverage of Unicode Basic Multilingual
                 Plane scripts for a GNU Unifont hex glyph file

   Synopsis: unicoverage [-ifont_file.hex] [-ocoverage_file.txt]

   This program requires the file "coverage.dat" to be present
   in the directory from which it is run.

   Author: Paul Hardy, unifoundry <at> unifoundry.com, 6 January 2008
   
   Copyright (C) 2008, 2013 Paul Hardy

   LICENSE:

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 2 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAXBUF 256


int
main (int argc, char *argv[])
{

   unsigned i;                /* loop variable                       */
   unsigned slen;             /* string length of coverage file line */
   char inbuf[256];           /* input buffer                        */
   unsigned thischar;         /* the current character               */

   char *infile="", *outfile="";  /* names of input and output files        */
   FILE *infp, *outfp;        /* file pointers of input and output files    */
   FILE *coveragefp;          /* file pointer to coverage.dat file          */
   int cstart, cend;          /* current coverage start and end code points */
   char coverstring[MAXBUF];  /* description of current coverage range      */
   int nglyphs;               /* number of glyphs in this section           */
   int nextrange();           /* to get next range & name of Unicode glyphs */

   if ((coveragefp = fopen ("coverage.dat", "r")) == NULL) {
      fprintf (stderr, "\nError: data file \"coverage.dat\" not found.\n\n");
      exit (0);
   }

   if (argc > 1) {
      for (i = 1; i < argc; i++) {
         if (argv[i][0] == '-') {  /* this is an option argument */
            switch (argv[i][1]) {
               case 'i':  /* name of input file */
                  infile = &argv[i][2];
                  break;
               case 'o':  /* name of output file */
                  outfile = &argv[i][2];
                  break;
               default:   /* if unrecognized option, print list and exit */
                  fprintf (stderr, "\nSyntax:\n\n");
                  fprintf (stderr, "   %s -p<Unicode_Page> ", argv[0]);
                  fprintf (stderr, "-i<Input_File> -o<Output_File> -w\n\n");
                  fprintf (stderr, "\nExample:\n\n");
                  exit (1);
            }
         }
      }
   }
   /*
      Make sure we can open any I/O files that were specified before
      doing anything else.
   */
   if (strlen (infile) > 0) {
      if ((infp = fopen (infile, "r")) == NULL) {
         fprintf (stderr, "Error: can't open %s for input.\n", infile);
         exit (1);
      }
   }
   else {
      infp = stdin;
   }
   if (strlen (outfile) > 0) {
      if ((outfp = fopen (outfile, "w")) == NULL) {
         fprintf (stderr, "Error: can't open %s for output.\n", outfile);
         exit (1);
      }
   }
   else {
      outfp = stdout;
   }
   slen = nextrange (coveragefp, &cstart, &cend, coverstring);
   nglyphs = 0;
   /*
      Print header row.
   */
   fprintf (outfp, "Covered      Range       Script\n");
   fprintf (outfp, "-------      -----       ------\n\n");
   /*
      Read in the glyphs in the file
   */
   while (slen != 0 && fgets (inbuf, MAXBUF-1, infp) != NULL) {
      sscanf (inbuf, "%x", &thischar);
      while (cend < thischar && slen != 0) {
         /* print old range total */
         fprintf (outfp, " %5.1f%%  U+%04X..U+%04X  %s\n",
                 100.0*nglyphs/(1+cend-cstart), cstart, cend, coverstring);
         /* start new range total */
         slen = nextrange (coveragefp, &cstart, &cend, coverstring);
         /*
            Count Non-characters as existing for totals counts
         */
         nglyphs = 0;
         if (cstart <= 0xFDD0 && cend >= 0xFDEF && nglyphs == 0)
            nglyphs += 32;
         else if (cstart <= 0xFFFE && cend >= 0xFFFF && nglyphs == 0)
            nglyphs += 2;
      }
      /*
         If we read in a noncharacter, don't count it -- we already
         counted it once above.
      */
      if (thischar < 0xFDD0 || (thischar > 0xFDEF && thischar < 0xFFFE))
         nglyphs++;
   }
   /* print last range total */
   fprintf (outfp, " %5.1f%%  U+%04X..U+%04X  %s\n",
           100.0*nglyphs/(1+cend-cstart), cstart, cend, coverstring);
   exit (0);
}

/*
   nextrange - get next Unicode range
*/
int
nextrange (FILE *coveragefp,
              int *cstart, int *cend,
              char *coverstring)
{

   int i;
   static char inbuf[MAXBUF];
   int retval;         /* the return value */

   if (fgets (inbuf, MAXBUF-1, coveragefp) != NULL) {
      retval = strlen (inbuf);
      if ((inbuf[0] >= '0' && inbuf[0] <= '9') ||
          (inbuf[0] >= 'A' && inbuf[0] <= 'F') ||
          (inbuf[0] >= 'a' && inbuf[0] <= 'f')) {
         sscanf (inbuf, "%x-%x", cstart, cend);
         i = 0;
         while (inbuf[i] != ' ') i++;  /* find first blank */
         while (inbuf[i] == ' ') i++;  /* find next non-blank */
         strcpy (coverstring, &inbuf[i]);
      }
   }
   else {
      retval = 0;
   }

   return (retval);
}
